### PR TITLE
Convert to Spring Boot and Three Layer Architecture

### PR DESCRIPTION
**Pull request by AelinaXY**
First iteration of Three Layer Architecture. TramStopService has also been refactored.

### CODE REVIEW

**Comment by AelinaXY** [comment]
The main things that have changed are the overall structure, addition of Spring Boot, and the refactor of TramStopService.update().

File: src/main/java/com/tfgm/TfgmApplication.java [hunk 1]
```diff
@@ -5,20 +5,22 @@
...
 
-public final class App {
-  private App() {}
+@SpringBootApplication
+public class TfgmApplication {
 
   public static void main(String[] args)
       throws URISyntaxException, IOException, InterruptedException {
-
+    //    SpringApplication.run(TfgmApplication.class, args);
```
Comment (thomas-ayling) [inline src/main/java/com/tfgm/TfgmApplication.java]: This is better, but you should use this line as it will take care of running the program for you.

Comment (AelinaXY) [reply src/main/java/com/tfgm/TfgmApplication.java]: It currently just breaks as it isn't connected to any db

File: src/main/java/com/tfgm/controllers/TramController.java [hunk 2]
```diff
@@ -1,163 +1,7 @@
 package com.tfgm.controllers;
```
Comment (thomas-ayling) [inline src/main/java/com/tfgm/controllers/TramController.java:1]: So now have a look at defining controllers and endpoints with Spring and start using that for your data output instead of printing. Your service should be auto wired in here and when you send requests to your endpoints call the service's methods.

File: src/main/java/com/tfgm/models/TramStop.java [hunk 3]
```diff
@@ -5,75 +5,33 @@
 import java.util.LinkedList;
```
Comment (thomas-ayling) [inline src/main/java/com/tfgm/models/TramStop.java]: Looks good, but why did you remove the JavaDocs?

File: src/main/java/com/tfgm/services/TramStopGraphService.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        if (findEndOfLine(
+            Arrays.stream(departingTram.getDestination().split(" "))
+                    .reduce((first, second) -> second)
+                    .get()
+                    .matches("MCUK|MediaCityUK")
+                ? "MediaCityUK"
+                : departingTram.getDestination(),
+            tramStopContainer.getTramStop())) {
+          tramStopContainer.getTramLinkStop().addTram(departingTram);
+          System.out.println(
```
Comment (thomas-ayling) [inline src/main/java/com/tfgm/services/TramStopGraphService.java:40]: This looks better, so now instead of printing the result return it as a String and then in your controller you can return the data as a ResponseEntity to the client.

File: src/main/java/com/tfgm/services/TramStopService.java [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+              .toArray(TramStopContainer[]::new),
+          tempStop.getJSONArray("nextStop").toList().stream()
+              .map(tramStopHashMap::get)
+              .map(TramStopContainer::new)
+              .toArray(TramStopContainer[]::new));
+    }
+
+    for (TramStop tramStop : tramStopHashMap.values()) {
+
+      System.out.println(tramStop.toString());
```
Comment (thomas-ayling) [inline src/main/java/com/tfgm/services/TramStopService.java]: Same here, just return the response as a String and send to client in controller

File: src/main/java/com/tfgm/services/TramStopService.java [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+
+    URIBuilder builder = new URIBuilder("https://api.tfgm.com/odata/Metrolinks");
+    // "StationLocation":"Withington","AtcoCode":"9400ZZMAWIT2","Direction":"Outgoing","Dest0":"East
+    // Didsbury","Carriages0":"Single","Status0":"Due","Wait0":"3","Dest1":"East
+    // Didsbury","Carriages1":"Double","Status1":"Due","Wait1":"14","Dest2":"East
+    // Didsbury","Carriages2":"Single","Status2":"Due","Wait2":"16","Dest3":"","Carriages3":"","Status3":"","MessageBoard":"Welcome to Metrolink. For up to date travel information contact Metrolink twitter on @MCRMetrolink or visit www.TfGM.com. Take care of each other Manchester.","Wait3":"","LastUpdated":"2023-05-26T15:29:19Z"
+
+    URI uri = builder.build();
+    HttpGet request = new HttpGet(uri);
+    request.setHeader("Ocp-Apim-Subscription-Key", "810dfa20bbd44815b07d8cb6f9d3be96");
```
Comment (thomas-ayling) [inline src/main/java/com/tfgm/services/TramStopService.java:61]: So this acts as your data layer, as it's where you're requesting data from.Add this logic into the persistence layer and add your data manipulation logic in there, then return the data in the correct format back to your service layer where you can perform business logic. 

Client request -> Controller -> Service -> Persistence (Manipulate data to be ready for service layer) ->  Service (Perform business logic) -> Controller -> Client response

File: src/main/java/com/tfgm/services/TramStopService.java [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+      for (TramStopContainer n : i.getPrevStops()) {
+        System.out.println(n.getTramLinkStop());
+      }
+    }
+  }
+
+  public void update() throws URISyntaxException, IOException {
+    HttpClient httpclient = HttpClients.createDefault();
+
+    URIBuilder builder = new URIBuilder("https://api.tfgm.com/odata/Metrolinks");
```
Comment (thomas-ayling) [inline src/main/java/com/tfgm/services/TramStopService.java:48]: Consider adding variables such as URIs to your application.properties for easy access

File: src/main/java/com/tfgm/models/TramStop.java (hunk 3 above)
Comment (AelinaXY) [reply src/main/java/com/tfgm/models/TramStop.java]: unsure, added them back in now

File: src/main/java/com/tfgm/TfgmApplication.java [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+  public static void main(String[] args)
+      throws URISyntaxException, IOException, InterruptedException {
+//    SpringApplication.run(TfgmApplication.class, args);
+
+      mainRunner();
+
+  }
+
+  @EventListener(ApplicationReadyEvent.class)
```
Comment (thomas-ayling) [inline src/main/java/com/tfgm/TfgmApplication.java:27]: Have a look at https://www.baeldung.com/spring-boot-configure-multiple-datasources for configuring an existing API as your datasource, this should let the application run as expected

**Review by thomas-ayling — CHANGES_REQUESTED** [review]

**Comment by AelinaXY** [comment]
Add MongoDB calls and saving Tram Network to Mongo
